spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java

```diff
@@ -305,12 +351,21 @@ public void unlock() {
 			try {
 				while (true) {
 					try {
-						this.mutex.delete(this.path);
-						return;
+						if (this.mutex.delete(this.path)) {
+							return;
+						}
+						else {
+							throw new IllegalStateException();
+							// the lock is no longer owned by current process, the exception should be handle and rollback the execution result
+						}
 					}
 					catch (TransientDataAccessException | TransactionTimedOutException | TransactionSystemException e) {
 						// try again
 					}
+					catch (IllegalStateException e) {
+						throw new IllegalStateException("Lock was released in the store due to expiration. " +
```

Comment:
OK. I see why you use `boolean` for `delete()` now.
I wonder if we can use a `ConcurrentModificationException` instead.
Or something from the `org.springframework.dao` package, like `ConcurrencyFailureException`

Reply:
I see. Well, that was just a suggestion if that is makes sense to rely on some specific exception instead of this common.
Might be the fact that `ConcurrentModificationException` does not fit in our locks scenario.
However that `ConcurrencyFailureException` might be OK.

Reply:
```
catch (TransientDataAccessException | TransactionTimedOutException | TransactionSystemException e) {
      // try again
}
```
In the current implementation, `TransientDataAccessException`, `TransactionTimedOutException`, and `TransactionSystemException` will be caught, and then the `delete()` method will be invoked again.

Since `ConcurrencyFailureException` extends `TransientDataAccessException`, I think we should not throw it for this scenario.

I will change it to throw `ConcurrentModificationException` instead if you think `ConcurrentModificationException` fits better.

Reply:
I see. Thanks. Missed that.

So, yeah, this `ConcurrentModificationException` fits in my brain because its name means exactly what happens here for us.
But I'm not sure if this is a correct scenario where we can use it from Java perspective.
That's why I'm reaching you for other opinion, arguments and so on.
Just want to be sure that we are on the same page with a decision to move on.

Reply:
I see, 
I agree that changing from `IllegalStateException` to `ConcurrentModificationException` can improve the readability.
But I'm also not sure if this is a correct scenario where we can use it from Java perspective.

How about we create our own exception -`LockOwnershipExpiredException` for Spring distributed locks?
It can be used by both `RedisLock`, and `JdbcLock`.